### ISSUE TITLE
Use best model in Trial during hyperparameter search

### ISSUE THREAD

**Issue by antonioalegria** [huggingface/transformers#25247]
### Feature request

When running a `Trainer.hyperparameter_search`, each trial's value is calculated from the last epoch's chosen metric. However, especially when using early stopping and `load_best_model_at_end`, it would be useful to use the best model instead.

This could be a parameter of `Trainer.hyperparameter_search` or a an overridable function getting the best value, or some callback.

### Motivation

Often, we use early stopping and take the best model from a particular run because it's possible for models to start overfitting and dropping off after a certain number of epochs. This phenomenon can also appear during hyper parameter search and, as such, we'd like to be able to use the best epoch's value to compare trials.

Without this we may get results that are not fully representative.

### Your contribution

Happy to help testing or in other ways I can. Not sure where to start but if there is a clear place to do it I'd be open to help.

**Comment by amyeroberts** [comment]
cc @sgugger

**Comment by sgugger** [comment]
Yes this is not currently supported. Could be nice to add, but this is not high-priority on our side, so it would have to be a contribution :-) Happy to review a PR!

**Comment by jasper-lu** [comment]
Hi, I was looking to contribute here so I dug into the code. @antonioalegria , what hyperparameter search backend are you searching? Based on my reading of the code, it seems this is already supported for the `RAY` backend. Passing in `ray_scope=all` to `TrainingArguments` should work. 
  * Code pointer on [huggingface](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/integration_utils.py#L360) end.
  * https://docs.ray.io/en/latest/tune/api/doc/ray.tune.ExperimentAnalysis.get_best_trial.html#ray-tune-experimentanalysis-get-best-trial

I can look to contribute this for the wandb and optuna backends. However, the sigopt backend seems rather opaque..
